storage/remote/write_handler.go

```diff
@@ -256,87 +269,178 @@ func (h *writeHandler) write(ctx context.Context, req *prompb.WriteRequest) (err
 	if samplesWithInvalidLabels > 0 {
 		h.samplesWithInvalidLabelsTotal.Add(float64(samplesWithInvalidLabels))
 	}
-
 	return nil
 }
 
-func (h *writeHandler) writeV2(ctx context.Context, req *writev2.Request) (err error) {
-	outOfOrderExemplarErrs := 0
+const (
+	prw20ReceivedSamplesHeader   = "X-Prometheus-Remote-Write-Received-Samples"
+	rw20ReceivedHistogramsHeader = "X-Prometheus-Remote-Write-Received-Histograms"
+	rw20ReceivedExemplarsHeader  = "X-Prometheus-Remote-Write-Received-Exemplars"
+)
+
+type responseStats struct {
+	samples    int
+	histograms int
+	exemplars  int
+}
 
-	timeLimitApp := &timeLimitAppender{
+func (s responseStats) SetResponseHeaders(h http.Header) {
+	h.Set(prw20ReceivedSamplesHeader, strconv.Itoa(s.samples))
+	h.Set(rw20ReceivedHistogramsHeader, strconv.Itoa(s.histograms))
+	h.Set(rw20ReceivedExemplarsHeader, strconv.Itoa(s.exemplars))
+}
+
+// writeV2 is similar to write, but it works with v2 proto message,
+// allows partial 4xx writes and gathers statistics.
+//
+// writeV2 returns the statistics.
+// In error cases, writeV2, also returns statistics, but also the error that
+// should be propagated to the remote write sender and httpCode to use for status.
+//
+// NOTE(bwplotka): TSDB storage is NOT idempotent, so we don't allow "partial retry-able" errors.
+// Once we have 5xx type of error, we immediately stop and rollback all appends.
+func (h *writeHandler) writeV2(ctx context.Context, req *writev2.Request) (_ responseStats, _ error, httpCode int) {
+	app := &timeLimitAppender{
 		Appender: h.appendable.Appender(ctx),
 		maxTime:  timestamp.FromTime(time.Now().Add(maxAheadTime)),
 	}
 
-	defer func() {
-		if err != nil {
-			_ = timeLimitApp.Rollback()
-			return
+	rs := responseStats{}
+	if err, httpCode := h.appendV2(app, req, &rs); err != nil {
+		if httpCode/5 == 100 {
+			// On 5xx, we always rollback, because we expect
+			// sender to retry and TSDB is not idempotent.
+			if rerr := app.Rollback(); rerr != nil {
+				level.Error(h.logger).Log("msg", "writev2 rollback failed on retry-able error", "err", rerr)
+			}
```

Comment:
this is just a lack of knowledge on my part, but does `app.Rollback()` not also rollback samples in the appender that were collected via scrape but not yet comitted to tsdb?

Reply:
:+1: awesome, thanks for checking